egs/swbd/run.sh

```diff
@@ -57,6 +57,18 @@ for order in 3 4 5; do
     format_arpa_lm.py data/lm_20k_${order}_prune${threshold} | gzip -c > data/arpa/20k_${order}gram_prune${threshold}.arpa.gz
 
   done
+
+# example of pruning by size.
```

Comment:
Thanks for finding these.
I fixed the issue with the space that should have been tab.
The problems you experienced with prune_lm.py, we are aware of and Jian is
working on fixing it (it involves a lot of changes and it's currently being
tested).
I'll wait for your pull request to fix the remaining issue.
Dan



> Hello, Dan and Jian There are few possible bugs I believed i) the
> generated arpa file contains the line <s> (the line after \1-grams) is
> separated by "tab and space”, which should be “tab and tab” ii) in
> swbd/run.sh lm_dir="data/lm/" arpa_dir="data/arpa/“ it results print some
> thing like data/lm//20000_5.pocolm, which doesn’t really matter, but make
> it a bit un-beautiful These two are quite trivial and I will fix it
> shortly. iii) The third one I am not quite sure When I use prune_lm.py to
> prune lm-dir to certain target n-gram counts, for some threshold it works
> (from 1003378333 to 408157132), but for some threshold it fails (like from
> 1003378333 to 7598178). Since I use the exact same way of doing these two,
> it is kind of weird (I tried this several times). The following is the log:
> … Fixing lm-state total 6.64527e+06 -> 6.64618e+06 Fixing lm-state total
> 5.97697e+06 -> 5.97853e+06 Fixing lm-state total 4.62547e+06 -> 4.62753e+06
> float-counts-prune: logprob change per word was -0.777808 over 3.59103e+08
> words. float-counts-prune: aside from unigram there were 19595753 nonzero
> n-grams. Of these 11 were not pruned because they were shadowed by a
> higher-order n-gram, and 2338106 because they lead to an existing LM-state
> (according to <protected-counts-input>). Of the 17257636 n-grams eligible
> for pruning, 17257636 were actually pruned. Traceback (most recent call
> last): File "/export/a08/zhouyang/pocolm/scripts/prune_lm_dir.py", line
> 647, in <module> (threshold, iter) = FindThreshold() File
> "/export/a08/zhouyang/pocolm/scripts/prune_lm_dir.py", line 599, in
> FindThreshold log_threshold = ls.Estimate(math.log(args.target_num_ngrams))
> File "/export/a08/zhouyang/pocolm/scripts/prune_lm_dir.py", line 544, in
> Estimate alpha = (self.y1 - self.y0) / (self.x1 - self.x0)
> ZeroDivisionError: float division by zero … I believe we may need to be
> cautious when some this special case when self.x1 == self.x0 occur. Please
> let me know if I could provide more details or any help on fixing this bug
> best regards, Zhouyang
> … <#m_4710640430120309100_>
> On Aug 15, 2016, at 4:29 PM, Daniel Povey ***@***.***> wrote: In
> scripts/internal/prune_size_model.py: > + # the num-xgrams decreased on
> the previous iteration. > + assert prev_num_xgrams >= cur_num_xgrams > +
> prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) ** (1.0/3.0)
> > + baseline_decrease_factor = 1.0 / 1.5 > + # choose the factor from these
> two, that's closest to one. > + predicted_decrease_factor =
> max(prev_change_factor, baseline_decrease_factor) > + > + # the following
> gives us the predicted num-xgrams if we were > + # to prune again with the
> same threshold 'cur_threshold'. > + predicted_num_xgrams_if_repeat =
> predicted_decrease_factor * cur_num_xgrams > + assert next_threshold >=
> cur_threshold > + > + predicted_extra_factor = (next_threshold /
> cur_threshold) ** (self.xgrams_change_power) > + return
> predicted_num_xgrams_if_repeat * predicted_extra_factor > + > + def
> AdjustModelForOvershoot(self): I'm responding here to your post about the
> thing not converging right because I think it has to do with this. I notice
> some issues with the code.. Firstly, you are doing some things in this
> function that don't have anything to do with adjusting the model. You
> should have a function Backtrack() that will do the actual backtracking.
> Adjusting the model is a separate issue. For backtracking, if the two most
> recent thresholds were identical, then you should remove the two most
> recent elements from the history [i.e. backtrack twice]. If the history
> would then be empty, then the user-specified initial threshold was too high
> and you can exit with failure. Secondly, the way you are adjusting the
> model needs to be changed to ensure convergence. The model doesn't have
> just one parameter [the xgrams_change_power], it has three parameters,
> although I'll suggest to remove one of them. Your code currently reads like
> this: prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) **
> (1.0/3.0) baseline_decrease_factor = 1.0 / 1.5 I think it would be a good
> idea to get rid of baseline_decrease_factor entirely, as a simplification,
> so that there will be no 'max' expression. Be sure to rename variables
> appropriately. [if we kept the baseline_decrease_factor, we'd have to
> update it when experiencing overshoot, by increasing it.] Have a variable
> self.prev_change_power = 0.5 which will replace the 1.0 / 3.0 power in your
> existing code. And when you adjust the model, replace your code
> self.xgrams_change_power *= float(cur_modeled_num_xgrams) / cur_num_xgrams
> with the following: self.xgrams_change_power *= 1.2 self.prev_change_power
> *= 1.2 if (self.prev_change_power > 1.0) self.prev_change_power = 1.0 The
> problem you are experiencing in your comment here #45
> <https://github.com/danpovey/pocolm/pull/45> (comment) is that at the
> point of failure, the xgrams_change_power is not the important part of the
> model, the important part is the 'predicted_decrease_factor', which is
> controlled by other parameters, which you are not updating. Next, the way
> you are adjusting the model is not very good. You need to adjust the — You
> are receiving this because you are subscribed to this thread. Reply to this
> email directly, view it on GitHub, or mute the thread.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/31/files/cc94865347ad60c12aa8db89f2c19959f2884edd#r75007646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu-0tSI0OsP6udvooIC4flNxiYnt8ks5qghWngaJpZM4JU463>
> .
>